File: leetcode/二叉树/diameter.ts

// Definition for a binary tree node.
class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}


function diameterOfBinaryTree(root: TreeNode | null): number {

    interface feature {
        diameter: number;
        maxDepth: number;
    }

    function recursive(root: TreeNode | null): feature {
        // return [diameter, maxDepth]
        if (!root) {
            return {diameter: 0 ,maxDepth: 0};
        }
        
        const left = recursive(root.left);
        const right = recursive(root.right);
        const curMaxDepth = Math.max(left.maxDepth, right.maxDepth);
        const dia1 = left.maxDepth + right.maxDepth;
        return {diameter: Math.max(left.diameter, right.diameter,dia1), maxDepth: curMaxDepth};
    }

    return recursive(root).diameter;

};

File: leetcode/二叉树/diameter.py
# Definition for a binary tree node.
from typing import Optional, Tuple

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        
        def recursive(root: Optional[TreeNode]) -> Tuple[int, int]:
            if not root:
                return (0, 0)
            left_d, left_max_depth = recursive(root.left)
            right_d, right_max_depth = recursive(root.right)

            max_depth = max(left_max_depth, right_max_depth) + 1

            # 如果经过 root，则 diameter 必然是 d1；
            # 否则，只可能是 d2 和 d3 
            d1 = left_max_depth + right_max_depth
            d2 = left_d
            d3 = right_d
            diameter = max(d1, d2, d3)
            return (diameter, max_depth)
        
        return recursive(root)[0]